include/rocksdb/table.h

```diff
@@ -51,6 +51,54 @@ enum ChecksumType : char {
   kxxHash64 = 0x3,
 };
 
+// `PinningTier` is used to specify which tier of block-based tables should
+// be affected by a block cache pinning setting (see `CachePinningOptinos`
+// below).
+enum class PinningTier {
+  // For compatibility, this value specifies to fallback to the behavior
+  // indicated by the deprecated options,
+  // `pin_l0_filter_and_index_blocks_in_cache` and
+  // `pin_top_level_index_and_filter`. When no overrides occur from those
+  // options (see their API documentation for exact overrides), `kNone` is used.
+  kFallback,
+
+  // This tier contains no block-based tables.
+  kNone,
+
+  // This tier contains block-based tables that may have originated from a
+  // memtable flush. In particular, it includes tables from L0 that are smaller
+  // than 1.5 times the current `write_buffer_size`. Note these criteria imply
+  // it can include intra-L0 compaction outputs and ingested files, as long as
+  // they are not abnormally large for L0.
+  kMaybeFlushed,
+
+  // This tier contains all block-based tables.
+  kAll,
+};
+
+// `CachePinningOptions` contains members indicating the desired block cache
+// pinning behavior for the different categories of blocks.
+struct CachePinningOptions {
+  // The tier of block-based tables whose top-level index of partitioned
+  // meta-blocks will be pinned.
+  //
+  // Note `cache_index_and_filter_blocks` must be true for this option to have
+  // any effect. Otherwise the top-level index of partitioned meta-blocks would
+  // be held in table reader memory, outside the block cache.
+  PinningTier metablock_top_level_index_pinning = PinningTier::kFallback;
```

Comment:
Carried over unaddressed comment from https://github.com/ajkr/rocksdb/commit/fe6ed27b4281e7275c31b6ca89f417387612c684:

> Please clarify what you mean by "meta-block" as this is not a standard term in public API. And is "metablock" redundant in these fields? Would we consider pinning something that's not a metadata block?

Reply:
I plan to rename the members `partition_pinning`, `top_level_index_pinning`, and `unpartitioned_pinning`? Since metadata is redundant, we can rename the structure `MetadataCachePinningOptions`.